### PR TITLE
iptag changes for reservedip

### PR DESCRIPTION
**Pull request by skomarag**
## Description

1. Adding new field called IPTag as part of RReservedIpContext in ServiceMangement/Network
2. Adding tests to test the new Field 
3. New command NewAzureIpTag that takes 2 values, type and Value

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by markcowl** [comment]
@skomarag test assembly not compiling

File: src/ServiceManagement/Compute/Commands.ServiceManagement.Test/Properties/Resource.Designer.cs [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
 //     Changes to this file may cause incorrect behavior and will be lost if
 //     the code is regenerated.
 // </auto-generated>
 //------------------------------------------------------------------------------
 
-namespace Microsoft.WindowsAzure.Commands.ServiceManagement.Test.Properties
-{
-
-
+namespace Microsoft.WindowsAzure.Commands.ServiceManagement.Test.Properties {
```
Comment (cormacpayne) [inline src/ServiceManagement/Compute/Commands.ServiceManagement.Test/Properties/Resource.Designer.cs]: @skomarag are the changes to this file necessary?

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Network/NewAzureIPTag.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                WriteObject(new IPTag { IPTagType = Management.Network.Models.IPTagType.FirstPartyUsage, Value = Value }, true);
+            }
+            else if (string.Equals(Management.Network.Models.IPTagType.AvailabilityZone, IPTagType, StringComparison.OrdinalIgnoreCase))
+            {
+                WriteObject(new IPTag { IPTagType = Management.Network.Models.IPTagType.AvailabilityZone, Value = Value }, true);
+            }
+            else
+            {
+                throw new Exception(IPTagType + Resources.InvalidIPTagTypeException);
```
Comment (cormacpayne) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Network/NewAzureIPTag.cs]: @skomarag this should never be possible due to the `ValidateSet` attribute you have on the `IPTagType` parameter

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Network/NewAzureReservedIP.cs [hunk 3]
```diff
@@ -93,6 +94,14 @@ public string Slot
             set;
         }
 
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, ParameterSetName = ReserveNewIPParamSet, HelpMessage = "List of IPTags.")]
+        [ValidateNotNullOrEmpty]
+        public List<IPTag> IPTags
```
Comment (cormacpayne) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Network/NewAzureReservedIP.cs]: @skomarag please rename this parameter to `IPTag` - having a singular noun for the parameter name follows the [PowerShell strongly encouraged developer guidelines](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#singular-noun-1)

File: src/ServiceManagement/Compute/Commands.ServiceManagement/Microsoft.WindowsAzure.Commands.ServiceManagement.dll-Help.xml [hunk 4]
```diff
@@ -29794,6 +29794,36 @@ $backupConfig=New-AzureVMSqlServerAutoBackupConfig -Enable -ResourceGroupName cr
...
+</command:syntax>
+<command:parameters></command:parameters>
+<command:inputTypes></command:inputTypes>
+<command:returnValues></command:returnValues>
+<maml:alertSet><maml:alert><maml:para>
+</maml:para>
+</maml:alert>
+</maml:alertSet>
+<command:examples><command:example><maml:title>Example 1</maml:title>
+<dev:code>PS C:\&gt; {{ Add example code here }}</dev:code>
```
Comment (cormacpayne) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/Microsoft.WindowsAzure.Commands.ServiceManagement.dll-Help.xml]: @skomarag please fill out the example and corresponding description for this cmdlet's help content

File: src/ServiceManagement/Compute/Commands.ServiceManagement/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -28,8 +28,8 @@
 // You can specify all the values or you can default the Build and Revision Numbers 
 // by using the '*' as shown below:
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyVersion("4.0.0")]
-[assembly: AssemblyFileVersion("4.0.0")]
+[assembly: AssemblyVersion("3.5.0")]
```
Comment (cormacpayne) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/Properties/AssemblyInfo.cs]: @skomarag please revert the changes made to this file

File: src/ServiceManagement/Compute/Commands.ServiceManagement.Test/Properties/Resource.Designer.cs (hunk 1 above)
Comment (skomarag) [inline src/ServiceManagement/Compute/Commands.ServiceManagement.Test/Properties/Resource.Designer.cs]: This is an autogenerated file

File: src/ServiceManagement/Compute/Commands.ServiceManagement/Microsoft.WindowsAzure.Commands.ServiceManagement.dll-Help.xml (hunk 4 above)
Comment (skomarag) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/Microsoft.WindowsAzure.Commands.ServiceManagement.dll-Help.xml]: ok

Comment (skomarag) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/Microsoft.WindowsAzure.Commands.ServiceManagement.dll-Help.xml]: Done

File: src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1 [hunk 6]
```diff
@@ -912,6 +912,30 @@ function Test-MigrationAzureReservedIP
...
 
+<#
+.SYNOPSIS
+Tests New-AzureReservedIPWithTags
+#>
+function Test-AzureReservedIPWithIPTags
+{
+    # Setu
+    $name = getAssetName
+    $location = "West Central US"
```
Comment (markcowl) [inline src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1]: Do not use hard-coded locations in tests

File: src/ServiceManagement/Common/Commands.ScenarioTest/ServiceManagement/ScenarioTests.cs [hunk 7]
```diff
@@ -200,6 +200,15 @@ public void TestMigrationAzureReservedIP()
...
         [Trait(Category.AcceptanceType, Category.BVT)]
+        public void TestAzureReservedIPWithIPTags()
+        {
+            this.RunPowerShellTest("Test-AzureReservedIPWithIPTags");
+        }
+
+        [Fact]
+        [Trait(Category.Service, Category.ServiceManagement)]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        [Trait(Category.AcceptanceType, Category.BVT)]
```
Comment (markcowl) [inline src/ServiceManagement/Common/Commands.ScenarioTest/ServiceManagement/ScenarioTests.cs]: You only need the second trait (```[Trait(Category.AcceptanceType, Category.CheckIn)]```)  Please remove the others.

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Network/NewAzureIPTag.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Management.Automation;
+using Microsoft.WindowsAzure.Management.Network.Models;
+using Microsoft.WindowsAzure.Commands.ServiceManagement.Properties;
+
+namespace Microsoft.WindowsAzure.Commands.ServiceManagement.IaaS
+{
+    [Cmdlet(VerbsCommon.New, "AzureIPTag"), OutputType(typeof(IPTag))]
```
Comment (markcowl) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Network/NewAzureIPTag.cs]: Must support ShouldProcess

File: src/ServiceManagement/Network/Commands.Network.Test/ReservedIP/ReservedIPTest.cs [hunk 9]
```diff
@@ -401,5 +403,67 @@ public void SetAzureReservedIPAssociationMultivip()
             Assert.Equal(1, mockCommandRuntime.OutputPipeline.Count);
             Assert.Equal("Succeeded", ((ManagementOperationContext)mockCommandRuntime.OutputPipeline[0]).OperationStatus);
         }
+
+        [Fact]
+        [Trait(Category.Service, Category.Network)]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
```
Comment (markcowl) [inline src/ServiceManagement/Network/Commands.Network.Test/ReservedIP/ReservedIPTest.cs]: You only need the second trait

File: src/ServiceManagement/Network/Commands.Network.Test/ReservedIP/ReservedIPTest.cs [hunk 10]
```diff
@@ -401,5 +403,67 @@ public void SetAzureReservedIPAssociationMultivip()
...
+                            && p.IPTags == iptags),
+                    It.IsAny<CancellationToken>()),
+                Times.Once());
+
+            Assert.Equal("Succeeded", ((ManagementOperationContext)mockCommandRuntime.OutputPipeline[0]).OperationStatus);
+        }
+
+        [Fact]
+        [Trait(Category.Service, Category.Network)]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
```
Comment (markcowl) [inline src/ServiceManagement/Network/Commands.Network.Test/ReservedIP/ReservedIPTest.cs]: same comment

File: src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1 (hunk 6 above)
Comment (skomarag) [inline src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1]: HI Mark,

This feature is not enabled in all regions.. it is infact enabled only in 1 region that is West Central US

File: src/ServiceManagement/Common/Commands.ScenarioTest/ServiceManagement/ScenarioTests.cs (hunk 7 above)
Comment (skomarag) [inline src/ServiceManagement/Common/Commands.ScenarioTest/ServiceManagement/ScenarioTests.cs]: Done

File: src/ServiceManagement/Network/Commands.Network.Test/ReservedIP/ReservedIPTest.cs (hunk 9 above)
Comment (skomarag) [inline src/ServiceManagement/Network/Commands.Network.Test/ReservedIP/ReservedIPTest.cs]: Done

File: src/ServiceManagement/Network/Commands.Network.Test/ReservedIP/ReservedIPTest.cs (hunk 10 above)
Comment (skomarag) [inline src/ServiceManagement/Network/Commands.Network.Test/ReservedIP/ReservedIPTest.cs]: Done

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Network/NewAzureIPTag.cs (hunk 8 above)
Comment (skomarag) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Network/NewAzureIPTag.cs]: Added

File: src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1 (hunk 6 above)
Comment (markcowl) [inline src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1]: Yes, the issue is not the region you are using,. it is that you are using a hard-coded region.  You should get the region by querying for the available regions for this resource.  Without this, it is not possible to run this test against other azure instances (test cluster, sovereign clouds, etc.)

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Network/NewAzureIPTag.cs [hunk 11]
```diff
@@ -19,7 +19,7 @@
 
 namespace Microsoft.WindowsAzure.Commands.ServiceManagement.IaaS
 {
-    [Cmdlet(VerbsCommon.New, "AzureIPTag"), OutputType(typeof(IPTag))]
+    [Cmdlet(VerbsCommon.New, "AzureIPTag", SupportsShouldProcess = true), OutputType(typeof(IPTag))]
```
Comment (markcowl) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Network/NewAzureIPTag.cs]: This is not sufficient.  You actually have to implement ShouldProcess

Comment (cormacpayne) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Network/NewAzureIPTag.cs]: [Here is some documentation on implementing `ShouldProcess`](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess)

**Comment by markcowl** [comment]
on demand run here: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1651/
